parsl/dataflow/dflow.py

```diff
@@ -736,3 +736,51 @@ def load_checkpoints(self, checkpointDirs):
             raise BadCheckpoint("checkpointDirs expects a list of checkpoints")
 
         return self._load_checkpoints(checkpointDirs)
+
+
+class DataFlowKernelLoader(object):
+    """Manage which DataFlowKernel is active.
+
+    This is a singleton class containing only class methods. You should not
+    need to instantiate this class.
+    """
+
+    _dfk = None
+
+    @classmethod
+    def set_default(cls, config, **kwargs):
+        """If there is no active DataFlowKernel, load one.
+
+        Args:
+            - config (dict) : Configuration to load. This config will be passed
+              to a new DataFlowKernel instantiation which will be set as the active
+              DataFlowKernel.
+            - **kwargs : Arbitrary keyword arguments which are passed to the
+              DataFlowKernel constructor.
+
+        """
+        if cls._dfk is None:
+            cls.load(config, **kwargs)
```

Comment:
Don't silently ignore a load if this is already initialised: probably throw an exception - it's a horrible user experience to spend 3 days wondering why your configuration is being ignored.

Reply:
I don't want to throw an exception-- the point of this method is that it should only load if nobody else has loaded first. If you want to load no matter what, you should use `load` instead of `set_default`. Would more verbosity here be enough? ("Configuration has already been loaded; default will have no effect" and "Using default configuration").

Reply:
The specific failure case (that I have encountered on other projects) I am trying to prevent is:

A user, while hacking round on their application code, end up sticking in a bunch of inits in all different files.

Then six months later, when they've forget that they've done that, they modify one to include some different config options.

Then they spend a long time trying to understand why their config options aren't being applied.

I don't want people to load no matter what - I want to stop people putting multiple init calls in in the first place - or more specifically, specifying multiple configurations, and having one of them used based on the possibly fairly arbitrary choice of which of their modules was imported first.

At the very least output a warning, though I think it is insufficient.

Reply:
Understood. The reason I put this in in the first place was in order for `pytest` to have control over the DFK while still being able to run tests interactively. Now that the resolution is executed at calltime instead of declaration time, that isn't necessary, so I'll just remove this method, which I think should address your concern.